src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
  * actually transmitted matched a pre-computed size and hash
  */
 class Runner {
-  constructor(agent, agentOpts) {
-    this.agent = agent || new https.Agent(agentOpts || {}); 
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      agent: Joi.object(),
+      agentOpts: Joi.object(),
+      maxRetries: Joi.number().min(0).max(10).default(5),
+    }).without('agent', 'agentOpts').optionalKeys(['agent', 'agentOpts']));
+
+    this.agent = opts.agent || new https.Agent(opts.agentOpts || {}); 
+
+    this.maxRetries = opts.maxRetries;
+
   }
 
+  /**
+   * Run a request and retry if a retryable error occurs.  Only 500-series
+   * responses are considered retriable, so any other status code will return
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
+      current = this.maxRetries - 1;
+      followRedirect = false;
+    }
+
+    function sleepFor(n) {
+      return new Promise((resolve, reject) => {
+        let delay = Math.pow(2, n) * 100;
+        setTimeout(resolve, delay);
```

Comment:
This is very close to recommendations:
http://docs.aws.amazon.com/general/latest/gr/api-retries.html

I would suggest adding some jitter like we did in fast-azure-storage:
https://github.com/taskcluster/fast-azure-storage/blob/master/lib/utils.js#L99-L103
and taskcluster-client:
https://github.com/taskcluster/taskcluster-client/blob/master/lib/client.js#L370-L378

Since this is intended to be used in workers... I would be willing to set the `maxDelay` pretty high, and similar I would argue for a delay factor higher than 100ms. Considering that we have 2-3k of these machines latency of a few seconds extra in exchange for robustness seems acceptable.

I know delay factor > 100 ms is not inline with AWS recommendations.. But sleeping a little extra can't hurt.